### PR TITLE
Input refactor

### PR DESCRIPTION
**Pull request by daviddq**
Refactoring input-related stuff to the Input class. Some other improvements:

- "Spacebar" key works again to cycle-focus nodes
- "Alt+Spacebar" works again to cycle-edit nodes while on the editor
- Marquee keeps selecting when moving out of the browser window and dragging over the search, menu and quick action icons and stops selecting when mouseup is received, even when the mouse is outside of the browser.

### CODE REVIEW

File: src/js/classes/app.js [hunk 1]
```diff
@@ -81,6 +82,19 @@ export var App = function(name, version) {
...
 
+  // inEditor
+  //
+  // Indicates if we are in the editor view
+  this.inEditor = () => self.editing();
+
+  // inWorkspace
+  //
+  // Indicates if we are in the workspace view
+  this.inWorkspace = () => !self.editing();
```
Comment (blurymind) [inline src/js/classes/app.js:93]: not sure why do we need to abstract this with two separate functions that do the same

File: src/js/classes/app.js [hunk 2]
```diff
@@ -1528,6 +1072,7 @@ export var App = function(name, version) {
     return tagBeforeCursor;
   };
 
+  // TODO: seems to be unused. Delete
   this.testRunFrom = function(startTestNode) {
```
Comment (blurymind) [inline src/js/classes/app.js:1076]: yeah this is some legacy code from the old story tester that only worked on electron. Feel free to get rid

File: src/js/classes/input.js [hunk 3]
```diff
@@ -0,0 +1,351 @@
+export const Input = function(app) {
+  const self = this;
+
+  const MouseButton = Object.freeze({
+    'Left': 0,
+    'Middle': 1,
+    'Right': 2
+  });
+
+  const Key = Object.freeze({
```
Comment (blurymind) [inline src/js/classes/input.js:10]: Very nice :+1:  so tidy and organised, gives me joy :)

File: src/js/classes/input.js [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+    'X': 88,
+    'Y': 89,
+    'Z': 90
+  });
+
+  this.mouse = { x: 0, y:0 };
+  this.isDragging = false;
+  this.isScreenTouched = false;
+  this.isMiddleButtonDown = false;
+  this.isShiftDown = false;
```
Comment (blurymind) [inline src/js/classes/input.js:35]: would be nice to have this for other modifiers (ctrl, alt)

File: src/js/classes/app.js (hunk 1 above)
Comment (daviddq) [reply src/js/classes/app.js:93]: 1) For clarity: asking for app.inWorkspace() is clearer than asking for !self.editing()

2) In the near future there will be more cases like app.inStoryPlayer(), app.inSettingsDialog(). This provides a simple, unified, consistent and elegant way of knowing the app's state

File: src/js/classes/workspace.js [hunk 5]
```diff
@@ -14,6 +18,236 @@ export const Workspace = function(app) {
...
+  //
+  //
+  this.onPanLeft = function() {
+    app.transformOrigin[0] += self.getPanAmount();
+    app.translate(PAN_TRANSITION_TIME);
+  };
+
+  // onPanRight
+  //
+  //
```
Comment (blurymind) [inline src/js/classes/workspace.js:38]: not sure why the extra //, just adds more lines to scroll through

File: src/js/classes/input.js (hunk 4 above)
Comment (daviddq) [reply src/js/classes/input.js:35]: They will be present soon. For now I just limited the PR to moving some stuff to the Input class and fixing evident bugs.

**Comment by blurymind** [comment]
as per usual this is a fantastic refactor. I left a few minor review notes, otherwise the code looks good :+1: 
Will need to give this a test on desktop and smartphone when I get some time- to check for any regressions

File: src/js/classes/app.js (hunk 1 above)
Comment (blurymind) [reply src/js/classes/app.js:93]: ok, in that case this does sound like a sensible thing to do :)

File: src/js/classes/workspace.js (hunk 5 above)
Comment (daviddq) [reply src/js/classes/workspace.js:38]: 1) Every new method was supposed to have a brief description but at a certain point I got tired of writing comments. I'll complete that on an upcoming push.

2) I like some space between methods so it's easy to notice when one ends and a different one starts when scrolling. That's a personal preference and a proposal. If you find it disturbing I can remove it.

**Comment by daviddq** [comment]
This refactor needs a good QA on mobile devices. I develop and test on desktop, so maybe there are drawbacks on other platforms.

**Comment by blurymind** [comment]
> This refactor needs a good QA on mobile devices. I develop and test on desktop, so maybe there are drawbacks on other platforms.

If you have a smartphone, you can view yarn on localhost when its running, but you need to find out what your pc's broadcast ip is

File: src/js/classes/workspace.js (hunk 5 above)
Comment (daviddq) [reply src/js/classes/workspace.js]: I already completed the method descriptions.

**Comment by blurymind** [comment]
this refactor broke moving a selected node with touch input (tap once to select, tap+ drag to move, double tap to edit)

File: src/js/classes/app.js [hunk 6]
```diff
@@ -49,18 +54,14 @@ export var App = function(name, version) {
...
   this.editor = null;
   this.nodeVisitHistory = [];
-  this.mouseX = 0;
-  this.mouseY = 0;
   this.clipboard = '';
   this.nodeClipboard = [];
   this.speachInstance = null;
   this.selectedLanguageIndex = 6;
   this.language = null;
-  this.hasTouchScreen = false;
```
Comment (blurymind) [inline src/js/classes/app.js:63]: possible suspect of what broke the node drag with touch

File: src/js/classes/app.js [hunk 7]
```diff
@@ -89,9 +103,6 @@ export var App = function(name, version) {
     if (navigator.platform.indexOf('Linux') != -1) osName = 'Linux';
     if (navigator.platform.indexOf('Linux') != -1) osName = 'Linux';
     self.isElectron = navigator.userAgent.toLowerCase().includes('electron');
-    window.addEventListener('touchstart', function() {
```
Comment (blurymind) [inline src/js/classes/app.js:92]: you removed the touch start detection?

File: src/js/classes/input.js [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+          break;
+
+        case MouseButton.Middle:
+          app.workspace.onDragStart({ x: e.pageX, y: e.pageY });
+          break;
+        }
+      }
+    });
+
+    window.addEventListener('touchstart', () => {
```
Comment (blurymind) [inline src/js/classes/input.js:64]: oh wait, you  renamed it.

Preferably when you edit touch related code,please test it on a touch device.. your smartphone or tablet. Anything would do.

Thelocalhost IP of the dev server is

<broadcast IP (get it with ifconfig)>:8080

Comment (daviddq) [reply src/js/classes/input.js:64]: I'll try and fix...

Comment (daviddq) [reply src/js/classes/input.js:64]: Can you elaborate more on this? It's working on my phone.

✅ Test 1:
- Selected a node (tap)
- Moved it (tap *on* the node and drag)

✅ Test 2:
- Selected a node (tap)
- Moved it (tap *outside* the node and drag)

Comment (daviddq) [reply src/js/classes/input.js:64]: Nevermind. Now fails 🤷‍♂️

Comment (daviddq) [reply src/js/classes/input.js:64]: ¡Fixed! 🔨

**Comment by blurymind** [comment]
thanks for fixing this, I will give it a test tomorrow and merge if all works :) :+1:

**Comment by blurymind** [comment]
Looks good :+1: 
Thank you for working on this :)

merged! :tada:

**Pull request merged**